query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code, unused_imports)]
+
+/// ! A ConjoiningClauses (CC) is a collection of clauses that are combined with JOIN.
+/// ! The topmost form in a query is a ConjoiningClauses.
+/// !
+/// !---------------------------------------------------------------------------------------
+/// ! - Ordinary pattern clauses turn into FROM parts and WHERE parts using :=.
+/// ! - Predicate clauses turn into the same, but with other functions.
+/// ! - `not` turns into NOT EXISTS with WHERE clauses inside the subquery to
+/// !   bind it to the outer variables, or adds simple WHERE clauses to the outer
+/// !   clause.
+/// ! - `not-join` is similar, but with explicit binding.
+/// ! - `or` turns into a collection of UNIONs inside a subquery, or a simple
+/// !   alternation.
+/// !   `or`'s documentation states that all clauses must include the same vars,
+/// !   but that's an over-simplification: all clauses must refer to the external
+/// !   unification vars.
+/// !   The entire UNION-set is JOINed to any surrounding expressions per the `rule-vars`
+/// !   clause, or the intersection of the vars in the two sides of the JOIN.
+/// !
+/// ! Not yet done:
+/// ! - Function clauses with bindings turn into:
+/// !   * Subqueries. Perhaps less efficient? Certainly clearer.
+/// !   * Projection expressions, if only used for output.
+/// !   * Inline expressions?
+/// !---------------------------------------------------------------------------------------
+/// !
+/// ! `from` is a list of [source alias] pairs, suitable for passing to honeysql.
+/// ! `bindings` is a map from var to qualified columns.
+/// ! `known-types` is a map from var to type keyword (e.g., :db.type/ref)
+/// ! `extracted-types` is a mapping, similar to `bindings`, but used to pull
```

Comment:
Yes, exactly. If you have a query like `[:find ?v :where [_ _ ?v]]` we need to project `all_datoms.v` to get the value, and `all_datoms.value_type_tag` so we know how to treat `15` when it arrives.